oak-jcr/src/test/java/org/apache/jackrabbit/oak/jcr/ProtectedPropertyTest.java

```diff
@@ -0,0 +1,376 @@
+        session.save();
+
+        try {
+            String testUuid = ref.getProperty("jcr:uuid").getString();
+            test.setProperty("jcr:uuid", testUuid);
+            // note this fails even though test hasn't be set to mix:referenceable
+            session.save();
+            fail("Attempt so set a UUID already in use should fail");
+        } catch (ConstraintViolationException ex) {
+            // expected
+        } finally {
+            test.remove();
+            ref.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredAfterSettingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        test.addMixin(NodeType.MIX_REFERENCEABLE);
+        session.save();
+        try {
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            fail("Setting jcr:uuid after adding mixin:referenceable should fail");
+        } catch (ConstraintViolationException ex) {
+            // expected
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredAfterRemovingMixin() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        test.addMixin(NodeType.MIX_REFERENCEABLE);
+        session.save();
+        try {
+            // check jcr:uuid is there
+            String prevUuid = test.getProperty("jcr:uuid").getString();
+            test.removeMixin(NodeType.MIX_REFERENCEABLE);
+            session.save();
+            // ist jcr:uuid gone now?
+            try {
+                String newUuid = test.getProperty("jcr:uuid").getString();
+                fail("jcr:uuid should be gone after removing the mixin type, was " + prevUuid + ", now is " + newUuid);
+            } catch (PathNotFoundException ex) {
+                // expected
+            }
+            String testUuid = UUID.randomUUID().toString();
+            test.setProperty("jcr:uuid", testUuid);
+            session.save();
+            test.addMixin(NodeType.MIX_REFERENCEABLE);
+            session.save();
+            assertEquals(testUuid, test.getProperty("jcr:uuid").getString());
+            Node check = session.getNodeByIdentifier(testUuid);
+            assertTrue(test.isSame(check));
+        } finally {
+            test.remove();
+            session.save();
+        }
+    }
+
+    @Test
+    public void jcrMixinReferenceableOnNtUnstructuredAfterRemovingMixinButDanglingReference() throws RepositoryException {
+        Node test = testNode.addNode(TEST_NODE_NAME, NodeType.NT_UNSTRUCTURED);
+        test.addMixin(NodeType.MIX_REFERENCEABLE);
+        session.save();
+        Node ref = testNode.addNode(TEST_NODE_NAME_REF, NodeType.NT_UNSTRUCTURED);
```

Comment:
Nitpick: Earlier the name ref was used for the referenced node, not the referencing one. So we might want to change that to avoid confusion.